infra: add docker containers

sbinet opened the pull request:
This CL adds docker images with `go` and LLVM/CLang installed.
- go-clang/base holds a `go` install under `/usr/lib/go` and points GOPATH at /go
- go-clang/build holds a LLVM install

Fixes go-clang/gen#100.

sbinet commented:
first stab.

- needs to integrate with travis-ci ?
- needs to update the docs.

sbinet commented:
@go-clang/dev ping?

zimmski commented:
Still here! Just really busy, sorry. The usual December madness :-( I will try to work through the go-clang TODO this weekend.

marxriedl commented:
Same here, sorry. Maybe I will have a chance to look at it today

marxriedl commented on README.md:
```diff
@@ -27,15 +27,15 @@ You want to contribute to go-clang? GREAT! If you are here because of a bug you
 This repository, [gen](https://github.com/go-clang/gen), holds the code to generate new bindings from headers of Clang's C API. These bindings are then bootstrapped using the [bootstrap](https://github.com/go-clang/bootstrap) repository. The `bootstrap` repository holds all basic files, like the CI configuration and a Makefile, as well as some additional code to make the bindings more complete an ...
 
-To ease the development process we have our own development environment based on [Vagrant](https://www.vagrantup.com/). The provided Vagrantfile executed in the root of the repository will setup an Ubuntu VM with our currently used Go version as well as Clang 3.4 and will set up everything that is needed to development and handle new versions of Clang.
+To ease the development process we have our own development environment based on [Docker](https://www.docker.com/). The provided `go-clang/build` `docker` image (located under `docks/build`) executed in the root of the repository will setup an Ubuntu image with our currently used Go version as well as Clang 3.4 and will set up everything that is needed to development and handle new versions of Cl ...
```
Will the vagrant setup stuff completely vanish? It doesn't seem that way. 
Do we want to maybe support both, vagrant and docker?

sbinet replied:
is it worth supporting 2 things?
what can we do with `vagrant` that we can't with a container virtualization approach like `Docker`?
(I am genuinely asking)

marxriedl replied:
Similar as me :D 
If we want to go Docker maybe we should drop the whole Vagrant stuff completely (including script etc)?

sbinet replied:
right.
when I started this PR, I wondered whether I couldn't re-use some of the `vagrant` scripts... (in the end, I just took them for inspiration)
but maybe just deleting them is less confusing.

sbinet replied:
should I just delete the `vagrant` scripts then?

marxriedl replied:
Sorry for not responding, yes, I believe we can get rid of all the vagrant stuff. 
I did hope that @zimmski would respond as well, but he seems to busy atm. 

Although I wonder how much the development of go-clang will change with Docker instead of vagrant.

marxriedl replied:
Specifically, connecting to a docker container is a little bit more inconvenient than connecting to  a vagrant vm.

zimmski replied:
Finally I have a big enough block to review this :-)
- I am all for eliminating Vagrant for Docker. The development environment is simple enough to fully rely on docker.
- If we are talking about removing `vagrant-bootstrap.sh` and `vagrant-environment.sh` that is fine with me.

marxriedl commented on README.md:
```diff
@@ -52,9 +52,7 @@ Every PR must be prepared using the following commands:
-make install
-make test
-make lint
+make pr
```
pr is short for?

sbinet replied:
pull request :)
I am not particularly attached to that name, though.

marxriedl replied:
Then i'm not a fan of this naming tbh.
Though I need this command for a pull request I won't exclusively use it for that and pr always has (for me at least) something to do with a version control system adding, staging, commiting etc.

sbinet replied:
fine by me. any suggestion?
`make test` is already taken.
`make check` is perhaps too vague (and perhaps easily confused with `test`)

marxriedl replied:
make test-lint | test_lint?

sbinet replied:
`make test-all` ?
(`test-lint` looks like we are testing `lint`...)

marxriedl replied:
`make test-all` would suit better, yes and i can't come up with reasonable other suggestions

sbinet commented:
ok. renamed `pr` into `test-all`.
I have also renamed the `go-clang/xyz` containers into `goclang/xyz` so they can be published on the hub.
I have created an organization named `goclang` (`go-clang` isn't a valid organization name).

sbinet commented:
ping?

marxriedl commented:
goclang is fine, the missing "-" shouldn't bother us.

zimmski commented on Makefile:
```diff
@@ -1,4 +1,4 @@
-.PHONY: all install install-dependencies install-tools lint test test-verbose
+.PHONY: all install install-dependencies install-tools lint test test-all test-verbose
```
Please add the docker target too

zimmski commented on Makefile:
```diff
@@ -32,3 +32,14 @@ test:
 test-verbose:
 	CGO_LDFLAGS="-L`llvm-config --libdir`" go test -timeout 60s -race -v ./...
+
+docker-images: docks/build/Dockerfile docks/base/Dockerfile
```
a.) Concerning the prerequisites, I do not think that we need that. "docker build" has its own cache mechanism.
b.) How about renaming "docker-images" to "docker-build-images"?

zimmski commented on Makefile:
```diff
@@ -32,3 +32,14 @@ test:
+
+docker-images: docks/build/Dockerfile docks/base/Dockerfile
+	docker build --rm -f ./docks/base/Dockerfile  --tag=goclang/base .
+	docker build --rm -f ./docks/build/Dockerfile --tag=goclang/build .
```
goclang/build can be used to work with Clang 3.4 right? I am wondering if we should tag them with the Clang version?

zimmski commented on Makefile:
```diff
@@ -32,3 +32,14 @@ test:
+	docker build --rm -f ./docks/base/Dockerfile  --tag=goclang/base .
+	docker build --rm -f ./docks/build/Dockerfile --tag=goclang/build .
+
+test-all:
```
Since this is a docker command I am wondering if this should get the "docker-" prefix as well?

zimmski commented on docks/base/Dockerfile:
```diff
@@ -0,0 +1,27 @@
+## goclang/base
```
How about renaming the folder "docks" to "docker"?

zimmski commented on docks/base/Dockerfile:
```diff
@@ -0,0 +1,27 @@
+## goclang/base
+## an ubuntu-based docker container with go.
+from ubuntu:trusty
```
Dockerfile commands should be uppercase by convention

zimmski commented on docks/base/Dockerfile:
```diff
@@ -0,0 +1,27 @@
+## goclang/base
+## an ubuntu-based docker container with go.
+from ubuntu:trusty
+maintainer go-clang-devs@github.com/go-clang
```
I am a little baffled if we have an official mail address for the group? Do you know if GitHub automatically adds one?

zimmski commented on docks/base/Dockerfile:
```diff
@@ -0,0 +1,27 @@
+env CODENAME trusty
+env GO_VERSION 1.4.3
+
+run apt-get update -y && \
```
Should we include `apt-get upgrade -y` (after update) and `apt-get autoremove -y`(before clean)?

zimmski commented on docks/base/Dockerfile:
```diff
@@ -0,0 +1,27 @@
+maintainer go-clang-devs@github.com/go-clang
+
+env CODENAME trusty
+env GO_VERSION 1.4.3
```
We are running on 1.7.1 now :-)

zimmski commented on docks/base/Dockerfile:
```diff
@@ -0,0 +1,27 @@
+	go version
+
+## setup for go-clang
+workdir ${GOPATH}/github.com/go-clang/gen
```
I learned that all used volumes should be added in the Dockerfile. What do you think?

zimmski commented on docks/build/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+from goclang/base
+maintainer go-clang-devs@github.com/go-clang
+
+env LLVM_VERSION 3.4
```
I think that the version should be set-able from the outside
```
ARG LLVM_VERSION
ENV LLVM_VERSION ${LLVM_VERSION:-3.4}
```

zimmski commented on docks/build/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+
+## setup LLVM and CLang (TODO: remove. see go-clang/design#3 go-clang/gen#102)
+run ln -s /usr/bin/llvm-config-${LLVM_VERSION} /usr/bin/llvm-config && \
+	ln -s /usr/lib/x86_64-linux-gnu/libclang-${LLVM_VERSION}.so \
```
To my surprise the libclang symlink is not needed see #133

zimmski commented on docks/build/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+		  /usr/lib/x86_64-linux-gnu/libclang.so
+
+## setup for go-clang
+workdir ${GOPATH}/github.com/go-clang/gen
```
I am wondering if this shouldn't be already set by the base image?

zimmski commented on docks/build/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+env LLVM_VERSION 3.4
+
+## add repositories
+run apt-get install -y software-properties-common && \
```
I know that this will look rather ugly but how about doing all the apt-* commands within one RUN?

zimmski commented on docks/build/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+## update and upgrade
+run apt-get update -y && \
+	apt-get -V upgrade -y && \
+	apt-get -V autoremove -y
```
Let's do autoremove before the `apt-get clean`

zimmski commented on docks/build/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+run apt-get -V install -y \
+	clang-${LLVM_VERSION} git libclang-${LLVM_VERSION}-dev \
+	llvm-${LLVM_VERSION}-tools make \
+	gcc g++
```
Maybe I am missing something but do we really need gcc and g++?

zimmski commented on docks/build/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+## install needed packages
+run apt-get -V install -y \
+	clang-${LLVM_VERSION} git libclang-${LLVM_VERSION}-dev \
+	llvm-${LLVM_VERSION}-tools make \
```
git and make could be added to the base image.

zimmski commented on README.md:
```diff
@@ -52,9 +52,7 @@ Every PR must be prepared using the following commands:
-make install
-make test
-make lint
+make test-all
```
I am wondering if we should add additional make targets to run things inside the container, or should we advice a contributor to open a bash using the build image?

zimmski commented:
@sbinet: Done with the review. Sorry for taking ages :-/

zimmski commented:
I close this PR because it has been stale for so long and we have an open issue which links to this PR.
